### PR TITLE
Fix validator skipping a batch each round by using itertools.islice

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #2637

Fix https://github.com/pytorch/torchtitan/issues/2636 but using islice instead of moving the check to the end of the loop.

**Safe to merge**

### CODE REVIEW

**Comment by fegin** [comment]
All ROCm CI are failing. Not related to this PR.

File: torchtitan/components/validate.py [hunk 1]
```diff
@@ -232,14 +233,16 @@ def validate(
...
 
+        config = cast(Validator.Config, self.config)
         accumulated_losses = []
         device_type = utils.device_type
         num_steps = 0
 
-        for input_dict, labels in self.validation_dataloader:
-            # pyrefly: ignore [missing-attribute, unsupported-operation]
-            if self.config.steps != -1 and num_steps >= self.config.steps:
-                break
```
Comment (tianyu-l) [inline torchtitan/components/validate.py:242]: I didn't get what it means by
> As a result, the validation set is reduced by one batch in each validation round.

In my mind, the proper behavior is that
**At each validation step, we build an entirely new validation_dataloader, because you would hope each validation step to have the same set of validation data.**

Is the bug that we are not loading validation from the beginning, but loading from where we left last validation step?

cc @albertoveneri

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/components/validate.py (hunk 1 above)
Comment (albertoveneri) [reply torchtitan/components/validate.py:242]: Thank you, @tianyu-l, for pointing this out. I thought this was a different issue and was preparing a separate issue/PR for it.

If I'm not mistaken, in other frameworks such as NeMo (https://github.com/NVIDIA-NeMo/NeMo) the validation dataset is not reset after each validation round, so the model sees different batches each time. For brevity, let's call it "iterative validation."

This PR correctly fixes the off-by-one issue in the iterative validation: https://github.com/pytorch/torchtitan/issues/2636.

However, I strongly prefer rebuilding the validation dataset from scratch at each validation round. That's exactly what I do in my experiments: I simply rebuild the validation dataloader each time `validate` is called. The overhead is usually negligible.

In the new issue, I was going to suggest adding a parameter to the config to decide whether the user wants to use the "iterative validation" or not.

Thank you again.

Comment (tianyu-l) [reply torchtitan/components/validate.py:242]: let's just (re)build the dataset at each validation step -- no need to add a config option

Comment (fegin) [reply torchtitan/components/validate.py:242]: If we are going to rebuild the dataset for each validation step, we can close this PR.
